File: src/01_data_structures/07_graph/01_adjacency_list/implementations/go/adjacency_list_test.go
package adjacency_list

import (
	"reflect"
	"testing"
)

func TestAdjacencyListInstantiation(t *testing.T) {
	graph := NewAdjacencyList()
	expected := &AdjacencyList{graph: map[string][]string{}}

	if !reflect.DeepEqual(graph, expected) {
		t.Errorf("Expected %v, got %v", expected, graph)
	}
}

func TestAddVertex(t *testing.T) {
	graph := NewAdjacencyList()

	graph.AddVertex("a")
	expected1 := &AdjacencyList{
		graph: map[string][]string{"a": {}},
	}

	if !reflect.DeepEqual(graph, expected1) {
		t.Errorf("Expected %v, got %v", expected1, graph)
	}

	graph.AddVertex("b")
	expected2 := &AdjacencyList{
		graph: map[string][]string{"a": {}, "b": {}},
	}
	if !reflect.DeepEqual(graph, expected2) {
		t.Errorf("Expected %v, got %v", expected2, graph)
	}

	graph.AddVertex("c")
	expected3 := &AdjacencyList{
		graph: map[string][]string{"a": {}, "b": {}, "c": {}},
	}
	if !reflect.DeepEqual(graph, expected3) {
		t.Errorf("Expected %v, got %v", expected3, graph)
	}
}

func TestAddEdge(t *testing.T) {
	t.Run("Adds an undirected edge", func(t *testing.T) {
		graph := NewAdjacencyList()
		graph.AddVertex("a").AddVertex("b")
		graph.AddEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{
				"a": {"b"},
				"b": {"a"},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Adds a directed edge", func(t *testing.T) {
		graph := NewAdjacencyList()
		graph.AddVertex("a").AddVertex("b")
		graph.AddEdge("a", "b", EdgeOptions{Directed: true})

		expected := &AdjacencyList{
			graph: map[string][]string{
				"a": {"b"},
				"b": {},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Adds an edge into an empty graph", func(t *testing.T) {
		graph := NewAdjacencyList()
		graph.AddEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Adds an edge into a graph missing one vertex", func(t *testing.T) {
		graph := NewAdjacencyList().AddVertex("a")
		graph.AddEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{
				"a": {},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})
}

func TestRemoveEdge(t *testing.T) {
	t.Run("Removes edge from empty graph", func(t *testing.T) {
		graph := NewAdjacencyList()
		graph.RemoveEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Removes edge from graph with vertices but no edges", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			AddVertex("b").
			RemoveEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{
				"a": {},
				"b": {},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Removes edge from graph with edges", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			AddVertex("b").
			AddEdge("a", "b", EdgeOptions{}).
			RemoveEdge("a", "b", EdgeOptions{})

		expected := &AdjacencyList{
			graph: map[string][]string{
				"a": {},
				"b": {},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})
}

func TestRemoveVertex(t *testing.T) {
	t.Run("Removes vertex from empty graph", func(t *testing.T) {
		graph := NewAdjacencyList()
		graph.RemoveVertex("a")

		expected := &AdjacencyList{
			graph: map[string][]string{},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Removes vertex from graph with one vertex", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			RemoveVertex("a")

		expected := &AdjacencyList{
			graph: map[string][]string{},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})

	t.Run("Removes vertex from graph with vertices and edges", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			AddVertex("b").
			AddEdge("a", "b", EdgeOptions{}).
			RemoveVertex("a")

		expected := &AdjacencyList{
			graph: map[string][]string{
				"b": {},
			},
		}

		if !reflect.DeepEqual(graph, expected) {
			t.Errorf("Expected %v, got %v", expected, graph)
		}
	})
}

func TestGetNeighbors(t *testing.T) {
	t.Run("Gets neighbors from graph vertex", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			AddVertex("b").
			AddEdge("a", "b", EdgeOptions{})

		neighbors := graph.GetNeighbors("a")
		expected := []string{"b"}

		if !reflect.DeepEqual(neighbors, expected) {
			t.Errorf("Expected %v, got %v", expected, neighbors)
		}
	})

	t.Run("Gets neighbors from graph vertex with multiple calls", func(t *testing.T) {
		graph := NewAdjacencyList().
			AddVertex("a").
			AddVertex("b").
			AddEdge("a", "b", EdgeOptions{})

		neighbors := graph.GetNeighbors("a")
		expected := []string{"b"}

		if !reflect.DeepEqual(neighbors, expected) {
			t.Errorf("Expected %v, got %v", expected, neighbors)
		}
	})
}

File: src/01_data_structures/07_graph/01_adjacency_list/implementations/ts/adjacency_list.test.ts
import { AdjacencyList } from "./adjacency_list";

describe("AdjacencyList()", () => {
  test("Is instantiated", async () => {
    const graph = new AdjacencyList();

    expect(graph).toEqual({ graph: {} });
  });
});

describe("addVertex()", () => {
  test("Adds a vertex", async () => {
    const graph = new AdjacencyList();

    graph.addVertex("a");
    expect(graph).toEqual({ graph: { a: [] } });

    graph.addVertex("a");
    expect(graph).toEqual({ graph: { a: [] } });

    graph.addVertex("b");
    expect(graph).toEqual({ graph: { a: [], b: [] } });
  });
});

describe("addEdge()", () => {
  test("Adds an undirected edge", async () => {
    const graph = new AdjacencyList();
    graph.addVertex("a").addVertex("b");
    graph.addEdge("a", "b");

    expect(graph).toEqual({
      graph: {
        a: [
          {
            value: "b",
            weight: undefined,
          },
        ],
        b: [
          {
            value: "a",
            weight: undefined,
          },
        ],
      },
    });
  });

  test("Adds a directed edge", async () => {
    const graph = new AdjacencyList();
    graph.addVertex("a").addVertex("b");
    graph.addEdge("a", "b", { directed: true });

    expect(graph).toEqual({
      graph: {
        a: [{ value: "b", weight: undefined }],
        b: [],
      },
    });
  });

  test("Adds an edge into an empty graph", async () => {
    const graph = new AdjacencyList();
    graph.addEdge("a", "b");

    expect(graph).toEqual({ graph: {} });
  });

  test("Adds an edge into a graph missing one vertex", async () => {
    const graph = new AdjacencyList().addVertex("a").addEdge("a", "b");

    expect(graph).toEqual({ graph: { a: [] } });
  });
});

describe("removeEdge()", () => {
  test("Removes edge from empty graph", async () => {
    const graph = new AdjacencyList().removeEdge("a", "b");

    expect(graph).toEqual({ graph: {} });
  });

  test("Removes edge from graph with vertices but no edges", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .removeEdge("a", "b");

    expect(graph).toEqual({ graph: { a: [], b: [] } });
  });

  test("Removes edge from graph with edges", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .addEdge("a", "b")
      .removeEdge("a", "b");

    expect(graph).toEqual({ graph: { a: [], b: [] } });
  });
});

describe("removeVertex()", () => {
  test("Removes vertex from empty graph", async () => {
    const graph = new AdjacencyList().removeVertex("a");

    expect(graph).toEqual({ graph: {} });
  });

  test("Removes vertex from graph with one vertex", async () => {
    const graph = new AdjacencyList().addVertex("a").removeVertex("a");

    expect(graph).toEqual({ graph: {} });
  });

  test("Removes vertex from graph with vertices and edges", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .addEdge("a", "b")
      .removeVertex("a");

    expect(graph).toEqual({ graph: { b: [] } });
  });
});

describe("getNeighbors()", () => {
  test("Gets neighbors from non-existing vertex", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .addEdge("a", "b");
    const neighbors = graph.getNeighbors("c");

    expect(neighbors).toEqual(null);
  });
  test("Gets neighbors from graph vertex", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .addEdge("a", "b");
    const neighbors = graph.getNeighbors("a");

    expect(neighbors).toEqual([{ value: "b", weight: undefined }]);
  });

  test("Gets neighbors from graph vertex", async () => {
    const graph = new AdjacencyList()
      .addVertex("a")
      .addVertex("b")
      .addEdge("a", "b");
    const neighbors = graph.getNeighbors("a");

    expect(neighbors).toEqual([{ value: "b", weight: undefined }]);
  });
});
